### DISCUSSION TITLE
Elastic subnets and the bazaar

### DISCUSSION THREAD

**Discussion by aphexmunky** [ava-labs/avalanchego#1765] [Ideas]
## The Cathedral

Subnets allow the creation of new networks to benefit from avalanche consensus and be supported by the large validator community that exists.

Elastic subnets being permissionless have been designed with similar parameters and reward mechanisms as the primary chain, in that the elastic subnet is created with the required incentivisation properties. While that doesn't prevent a subnet from creating it's own additional incentivisation's within the virtual machines, it does create a restriction in what kinds of assets can be used as the basis for a staking token. That restriction being assets that the p-chain has complete control over the supply.

To build an elastic subnet, you need an avalanche asset that you need to solve a distribution problem for.

## The Bazaar

There are many existing coins and tokens with very good and wide distribution already established. Projects people believe in with communities, memes, subcultures, exchange support, AMM liquidity, p2p trading.

These communities don't always really use the token. Hodling is participating.

The problems with using these coins and tokens as staking tokens on new networks is
- because the distribution is already established, a whale could easily disrupt and attack that new network. 
- because the supply mechanism already exists, you can't create new incentivisation parameters based on the underlying supply of it

I would like to suggest that we allow these assets to be registered on the pchain as a distinct asset type with unknowable supply properties. You wouldn't be able to create an elastic subnet with this asset type but you could create a different kind of permissionless subnet that uses the token as a access guard for validating the subnet. Since the supply and distribution is unknowable for the pchain, the consensus weighting should be driven by the stake weighting from the avax on the primary chain.

Incentivisation for running a validator then rests on the virtual machine creator. We could have a BTC.b based subnet running a bitcoin vm where the transaction fees are split between validators based on the proportion of BTC.b they have locked up. The safety and liveness of the network would depend on the avalanche validator community rather than whether a bitcoin whale had entered yet. There likely will still exist times where the safety and liveness is attacked on some of these subnets but is it really a decentralised network if that wasn't a possibility.

**Comment by Dorlanz** [comment]
This is really interesting - and importan- discussion to have. Couple of thoughts:

If the consensus weight is based on AVAX and rewards are weighted on the network specific token

Doesn't it mean is very cheap to attack the network as an AVAX whale?

Reason being, your AVAX stake won't necessarily lose value if it's a small subnet being attacked and the attacker doesn't have subnet tokens. So it would be the subnet token's holders that are  harmed by the attack and not the attacker, who can profit from both the attack and the subnet token price going down.

**Reply by aphexmunky** [reply]
I would agree that it would make it cheap to attack the network if you were already an AVAX whale - but maybe that's not a bad thing. AVAX holders have an interest in subnets thriving. Since these are permissionless networks, there may be subnets that AVAX Validators don't want operating. If someone created a decentralised marketplace as a subnet based on human trafficking - then yeah, i'd attack that. If someone creates a dogecoin subnet with both the bitcoin vm and an EVM running and i can earn fees, what's my incentive to attack? It's just like voting imo

**Reply by Dorlanz** [reply]
Ah, I see your point! Didn't understand fully at first. This could somewhat work for big non AVAX native tokens like BTC or Doge, where the alignment of AVAX holders with the network could be greater than that of the holders of the network's token.

It might not work for smaller tokens that are AVAX native, but I guess that's not the point here

**Comment by StephenButtolph** [comment]
I have a couple questions:

---

> ... allow these assets to be registered on the pchain as a distinct asset type with unknowable supply properties. ... you could create a ... permissionless subnet that uses the token as a access guard for validating the subnet. ... the consensus weighting should be driven by the stake weighting from the avax on the primary chain.

In this world I don't _really_ see _what_ the distinct asset type is doing. In general, the only reason we have these assets is to provide sybil resistance... But if the sampling distribution is based on AVAX what does the distinct asset do on the P-chain?

---

If we are trying to allow the VM to define the incentivization model, is there a reason that we shouldn't allow the VM to also define the sampling distribution?

If we implement this - we could have a VM specify the AVAX token as a sampling model (as this proposal recommends)... But they could also implement some alternative sampling mechanism (even a PoW based mechanism)

**Reply by aphexmunky** [reply]
I'll use BTC.b here in lieu of a better name for the specific subnet token since staking token is conflated.

By distinct asset type, i was merely trying to differentiate the assets on pchain as to create a new restriction on [TransformSubnetTx](https://github.com/ava-labs/avalanchego/blob/0ec52a9c6e5b879e367688db01bb10174d70b212/vms/platformvm/txs/transform_subnet_tx.go#L48-L50) so that c-chain assets couldn't be used on elastic subnets to affect supply.

I can see your point about the sybil resistance in the context of multiple subnets. (I shouldn't be able to validate multiple BTC.b subnets on the same stake) but i'm failing to realise the consequences in the context of sampling distribution. Is this a concern that liveness of the subnet could be attacked by a large AVAX stake and minimum amount of BTC.b? It seems reasonable to me that the largest stakers of AVAX get the most votes on what its preferred blocks are. The BTC.b portion is more a nothing-at-stake factor.

But yes, I think it would be very interesting to allow some creative expression and experimentation on VMs playing with custom sampling distributions and mechanisms.

**Reply by StephenButtolph** [reply]
I guess my concrete question would be:

What would be the difference on the P-chain between a subnet validator that locks 1 BTC.b verse a subnet validator that locks 2 BTC.b?

If there is no difference, then I'm not sure that I can see why locking BTC.b is needed at all.

**Reply by aphexmunky** [reply]
Ok, gotchu. I'll explain my naive view of how something like this might be implemented and then explain why I think having the token locked on the pchain provides more value that just warping the asset in.

bitcoin vm subnet:
- no forked state
- no coinbase reward for blocks beyond the fee aggregation
- transaction fees go to a special address that tracks validator distributions weighted on BTC.b stake. They can be awarded / exported once validator stake has concluded with uptime requirements
- empty blocks rejected
- similar setup to c-chain in that the native asset have to be imported and exported
- SPV nodes as to effectively acts as a bitcoin wallet compatible L2.

So greater BTC.b stake grants you a greater fee share. Greater AVAX stake gives you greater weighting in consensus.

The escalation of the network. The competition of validators for reward is rooted in BTC. The more BTC you lock up. The greater your share from the success of the network. The more BTC is unavailable to be sold in the market. It's a network a maxi can support because metcalfe's law (from an infrastructure/decentralisation pov not user) is based on their asset and not someone elses.

Now while it may be simpler from a VM implementation to just say the proposer allocates the transaction fee to an address they control and there's no BTC.b lock up required to run a validator. You're pinning the growth of that network on a network the users don't necessarily care about or have any belief in. You have no faith that the validators of your network aren't adversarial to it because they have no skin in the game.

If you believe that a subnet can be more valuable than the primary chain, you have to foster an environment where the rewards for accruing that asset are greater than the rewards for accruing your own asset.

Or maybe it comes down to something as simple as bag holder bias.

If you tell me that's unworkable from a consensus point of view, i'll happily concede the point - but hopefully i've conveyed some dynamics that i think would make that network more interesting. I'd love to hold some liquid staked btc.b 😍

**Reply by StephenButtolph** [reply]
> So greater BTC.b stake grants you a greater fee share.

Ah - this was the key point I was missing.

I'd like to pull this out into a few separate points/ideas (please correct anything that I misunderstood)

## My attempt at a summary

1. The P-chain's subnet staking mechanism is restrictive.
2. The rewards logic is parametrizable, but not general, as assets must behave in a manner understood by the P-chain.
3. The consensus weighting is pre-described to be PoS based on the reward asset.
4. The existing staking mechanism rewards validators proportionally to their consensus weight. Rewards should be able to be provided with a mechanism independent to their consensus weight.
5. If rewards were independent to consensus weight, we could simply use AVAX to weight consensus on a subnet.

## My thoughts on my own summary

(assuming I did understand correctly)

> 1. The P-chain's subnet staking mechanism is restrictive.

I agree, the existing subnet staking mechanism _is_ too restrictive and prevents innovative ideas.

> 2. The rewards logic is parametrizable, but not general, as assets must behave in a manner understood by the P-chain.

I dislike that staking on the P-chain requires locking all rewards that can be allocated for the lifetime of the subnet.
Having an asset that is not controlled by the creator of the subnet or an inflationary asset should be supported on the P-chain.
This support shouldn't be added explicitly. A mechanism should be introduced that allows a VM implementer to programmatically define arbitrary asset properties.

> 3. The consensus weighting is pre-described to be PoS based on the reward asset.

There is no technical reason why the reward asset must be the same asset that determines consensus weighting.
It certainly makes the economics more complex, as the value of these two assets will be somehow intertwined.
In fact, the consensus weighting doesn't necessarily need to be an on chain asset.
The consensus weighting just needs to be somehow calculated on chain.

Up to this point, I agree with everything.

> 4. The existing staking mechanism rewards validators proportionally to their consensus weight. Rewards should be able to be provided with a mechanism independent to their consensus weight.

I don't know that rewards _can_ be independent from consensus weight. Validators are being paid to secure the chain. A validator's consensus weight is the amount of responsibility they have. By separating these mechanisms, I'd be concerned that the network does not provide an incentive to secure itself.

> 5. If rewards were independent to consensus weight, we could simply use AVAX to weight consensus on a subnet.

Related to my point above, I don't know that rewards can truly be independent of the consensus weight... In my opinion, using AVAX as the consensus weight would actually _prevent_ a subnet from being more valuable than the primary network. Specifically, because the security of the subnet would always be tied to AVAX.

**Reply by aphexmunky** [reply]
> Related to my point above, I don't know that rewards can truly be independent of the consensus weight

I don't think they necessarily are independent. The security of the whole network, not just the subnet is already tied to AVAX. My assumption is that validators will secure the most profitable subnets available to them and attempt to minimise their own costs. As such, large AVAX validators already bear that increased responsibility. It becomes a question of, do they own enough BTC to make running the subnet profitable or not. That's going to be dynamic in the system, but in my mind if we establish the primary chain to be behaving cooperatively then it's possible to establish subsets would also behave cooperatively.

Are the uptimes independent? If a validator is participating in consensus on 2 networks, does the failure to respond to one constitute a failure to respond in the other?

**Reply by StephenButtolph** [reply]
> I don't think they necessarily are independent. The security of the whole network, not just the subnet is already tied to AVAX. ... As such, large AVAX validators already bear that increased responsibility.

That's fair. I guess my hope would be that we wouldn't hardcode this mechanism as a subnet option - but instead allow a VM to somehow specify it.

> Are the uptimes independent? If a validator is participating in consensus on 2 networks, does the failure to respond to one constitute a failure to respond in the other?

It isn't really possible for the validator set of subnet A to measure the participation of validators in subnet B. There is still more work to be done around subnet uptime tracking... But only the validators of the subnet are able to form an opinion of their own uptimes. Those opinions however, can then be communicated to other subnets.

**Reply by aphexmunky** [reply]
Yeah, I agree on avoiding the hardcoding and abstracting out to allow the VM to provide as much of the mechanism as possible. While it may be interesting to track participation at the staked or consensus asset ID level, it would likely lead to false results if avalanchego evolved into more like cluster orchestrator piece.

My preference would be that avalanche provides the most freedom on how you build your subnet. I'd rather things were tried and failed over being restrictive to known good solutions.

Thanks for indulging my ravings in any case
